src/wcl/rope.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <memory>
+
+namespace wcl {
+
+class rope_impl_base {
+ protected:
+  // On construction this has to be set
+  size_t length;
+
+  rope_impl_base(size_t len) : length(len) {}
+
+ public:
+  // methods that all RopeImpl share
+  virtual void write(std::ostream&) const = 0;
+  virtual std::string as_string() const = 0;
```

Comment:
can this just always be implemented using an std::stringstream? You can do that in rope for instance.